packages/cross_file/lib/src/x_file.dart

```diff
@@ -2,6 +2,140 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-export 'types/interface.dart'
-    if (dart.library.js_interop) 'types/html.dart'
-    if (dart.library.io) 'types/io.dart';
+// ignore_for_file: avoid_unused_constructor_parameters
+
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'legacy/legacy_factory.dart';
+
+/// The interface for a cross-platform "file", [XFile].
+///
+/// An `XFile` is a container that wraps a bunch of [bytes] with
+/// some optional (platform-dependent) metadata, like a [path], [mimeType]
+/// or [name].
+///
+/// Depending on the platform, the data can be backed by several different
+/// storages, like a `File` from `dart:io`, or a `Blob` from `package:web`.
+///
+/// This class is a very limited subset of dart:io [File], so all
+/// the methods should seem familiar.
+///
+/// Use the platform-specific factories to create `XFile`s with the most
+/// appropriate backend for your platform. See the [XFileFactory] class
+/// available for `web` and `native`.
+abstract interface class XFile {
+  @Deprecated('Use XFileFactory.fromPath from native/factory.dart.')
+  factory XFile(
+    String path, {
+    String? mimeType,
+    String? name,
+    int? length,
+    Uint8List? bytes,
+    DateTime? lastModified,
+  }) {
+    if (bytes != null) {
+      return XFileLegacyFactory.fromBytes(
+        bytes,
+        mimeType: mimeType,
+        name: name,
+        length: length,
+        lastModified: lastModified,
+        path: path,
+      );
+    }
+    return XFileLegacyFactory.fromPath(
+      path,
+      mimeType: mimeType,
+      name: name,
+      length: length,
+      bytes: bytes,
+      lastModified: lastModified,
+    );
+  }
+
+  @Deprecated('Use XFileFactory.fromBytes from native (or web)/factory.dart.')
+  factory XFile.fromData(
+    Uint8List bytes, {
+    String? mimeType,
+    String? name,
+    int? length,
+    DateTime? lastModified,
+    String? path,
+  }) {
+    return XFileLegacyFactory.fromBytes(
+      bytes,
+      mimeType: mimeType,
+      name: name,
+      length: length,
+      lastModified: lastModified,
+      path: path,
+    );
+  }
+
+  /// Get the path of the picked file.
+  ///
+  /// **Not all `XFile` instances have a `path`.**
```

Comment:
I think the changes in this package are sufficiently big to do a breaking change, with all the changes we might want.

I just did a non-breaking change "because I could", but as you said above, it might be "subtly breaking elsewhere".

How do you feel about making this a breaking change? There would be extra work in revisiting the consumers of this package and fixing those (which would also need a major version bump because they tend to re-export XFile as well :S)

I don't mind tackling that as well, LMK if you're okay with the change!

Reply:
> There would be extra work in revisiting the consumers of this package and fixing those (which would also need a major version bump because they tend to re-export XFile as well :S)

Yes, working that through the ecosystem could be kind of rough. I think since your version doesn't make any real structural compromises to make a probably-not-breaking change, I want to dig into usage a bit. My gut is that nobody is doing complex subclassing that would be affected by it, but I want to try to validate that against at least the public code I can find, and especially some popular packages. If it look like it'll be safe (e.g., nobody is ever subclassing any of the classes), then we can do this version and make things easier on everyone.

Reply:
There's some subclassing going on, most of it to add extra attributes/methods to the underlying XFile instance (they could wrap the underlying XFile and add stuff to it, or use extension methods I think?), one that could maybe migrated to one of the new web constructors, one used to store the file in "Hive", and one that is importing from src (this is illegal!)

* `extends XFile` ([search](https://github.com/search?q=%22extends+XFile%22+language%3ADart&type=code&l=Dart&p=1))
  * https://github.com/MixinNetwork/flutter-plugins/blob/b2beb15590f7b6c38e650ae501923b9bd3e7849c/packages/desktop_drop/lib/src/drop_item.dart#L5 (adds properties to XFile)
  * https://github.com/vidklopcic/flutter_persistent_socket/blob/1f7dfd216e9360c4d6bfb7aad8b29c0a897d6d22/lib/uploader/file_picker/file_picker_web.dart#L30 (`BlobXFile.fromFile`?)
  * https://github.com/slatei/asset_manager/blob/ba5b362b3c6ad26986b6448750d245d4823e3c17/lib/shared/hashed_file.dart#L5 (adds properties to XFile)
  * https://github.com/Wayaer/fl_assets_picker/blob/a66c525a9d9d937c1f31750b8b9f2d76513fa6ae/fl_image_picker/lib/src/extension.dart#L7 (adds properties to XFile)
  * https://github.com/diego78795/contact_app/blob/0d224aad5fb31353336569408133ef9951223003/lib/adapters/image_adapter.dart#L4 (extension method on XFile?)
  * https://github.com/correiarangel/camera_poc/blob/7d9e37549aa90001dfd97ed72bf65c37002c5dff/lib/core/shared/services/camera/library/z_xfile.dart#L3 (proxies XFile)
  * https://github.com/BreX900/cross_file_picker/blob/02dec6132bf5dacb020a6592d847d219d3aeded8/lib/src/x_file_stream.dart#L8 (imports from 'src')
  * https://github.com/monzir3bdo/task/blob/d2d4f906ed34bdaa6e92cc73a7b83512a1f66509/lib/features/products/data/models/photo_model.dart#L9 (could use the new interface)
* `implements XFile` ([search](https://github.com/search?q=%22implements+XFile%22+language%3ADart&type=code&l=Dart))
  * All seems to be test code.
* (extension ...) `on XFile` ([search](https://github.com/search?q=%22on+XFile%22+language%3ADart&type=code))
  * Everything should work (?)

---

I could create issues in their respective owner's repos to point them to this PR so they can see what's coming before we land it?

Reply:
That looks like we would break existing packages, unfortunately. Warning people in advance isn't enough for that, we'd need to version accordingly.

So I think the options are:
1. Do a breaking change and deal with the resulting pain
2. Make the new interface a super-interface of `XFileBase`, and let individual packages adopt that incrementally (although that would still be a breaking change for the packages that export it in their API).

The main drawback to 2 is that we need to change the name of the class everyone uses, and we would be stuck with the legacy bits for however long we want to support both.

Reply:
I think I prefer 1.

With the current implementation, the package is backwards-compatible for ourselves, so the transition for our packages should be relatively painless (and we could allow a version range).

(Unless we wan to do the nullability changes at the same time :))

Reply:
If we're going to do a major version change, that's our chance to fix nullability; if we don't do it then we probably never will.

Reply:
OK, I'll do all the incompatible changes in one go. Can you think of anything else? Maybe remove the `saveTo` method from the class, and make it a helper instead?